pkg/cloud/rgraph/exec/executor_parallel.go

```diff
@@ -140,6 +172,10 @@ func (ex *parallelExecutor) runAction(ctx context.Context, a Action) error {
 }
 
 func (ex *parallelExecutor) queueRunnableActions() {
+	if ex.pq.IsDone() {
```

Comment:
Is this required or an optimization?

If it's required, I'm concerned about if this is a race use to doing this outside of a lock.

Reply:
IsDone is locked by parallel_queue lock. It is thread safe and we don't have to acquire parallel_executor lock. 
This is to prevent task scheduling on Error when error strategy is StopOnError. https://screenshot.googleplex.com/9QxSKBMk46cCWdR

We need to check the queue because paralell_queue will finish (set state done) on error [link](https://github.com/GoogleCloudPlatform/k8s-cloud-provider/blob/e4456cb76f3897d74cd69320aa2f4d769c3ec1f8/pkg/cloud/rgraph/algo/parallel_queue.go#L183) and there is no other way to check that (because context is not canceled yet). If we don't check that we will try to schedule the task, but parallel_queue will not process it and we will lock the information about this task state because it will not be in pending array.

Alternative option is to return status from [func (q *ParallelQueue[T]) Add(item T) {](https://github.com/GoogleCloudPlatform/k8s-cloud-provider/blob/e4456cb76f3897d74cd69320aa2f4d769c3ec1f8/pkg/cloud/rgraph/algo/parallel_queue.go#L117-L118). This way we will know if task will be schedule for processing but I think that checking IsDone is a better way.

Reply:
I've been thinking about that, and you are right. 
I think that we should return status in paralelQueue:Add function. this way we will ne 100% sure if action will be scheduled by the queue or not.

Reply:
fixed PTAL